src/main/scala/workflow/NodeOptimizationRule.scala

```diff
@@ -5,96 +5,341 @@ import org.apache.spark.rdd.RDD
+   * or fits an optimizable estimator,
+   * and is not supposed to be executed on test data.
+   */
+  private def getInstructionsToOptimize(instructions: Seq[Instruction]): Set[Int] = {
+    instructions.zipWithIndex.collect {
+      case (TransformerApplyNode(tIndex, _), index)
+        if instructions(tIndex).isInstanceOf[OptimizableTransformer[_, _]] => index
+
+      case (EstimatorFitNode(estIndex, _), index)
+        if instructions(estIndex).isInstanceOf[OptimizableEstimator[_, _]] => index
+
+      case (EstimatorFitNode(estIndex, _), index)
+        if instructions(estIndex).isInstanceOf[OptimizableLabelEstimator[_, _, _]] => index
+
+    }.toSet -- WorkflowUtils.getChildren(Pipeline.SOURCE, instructions)
+  }
+
+  /**
+   * Execute the instruction at the given index, and output the new state of the optimization process
+   *
+   * @param optimizationState
+   * @param instruction
+   * @param index
+   * @return
+   */
+  private def executeInstruction(optimizationState: OptimizationState, instruction: Instruction, index: Int)
+  : OptimizationState = {
+    // Get the dependencies from the registers and run the instruction on them (caching RDD outputs)
+    val registers = optimizationState.registers
+    val deps = instruction.getDependencies.map(registers)
+    val instructionOutput = instruction.execute(deps) match {
+      case RDDOutput(rdd) => RDDOutput(rdd.cache())
+      case out => out
+    }
+    val newRegisters = registers + (index -> instructionOutput)
+
+    // If any of the dependencies is an RDD and has a numPerPartition, make sure to feed it forward
+    val numPerPartitionPerNode = optimizationState.numPerPartitionPerNode
+    val numPerPartition = instruction.getDependencies.collectFirst {
+      case dep if numPerPartitionPerNode.contains(dep) => numPerPartitionPerNode(dep)
+    }
+    val newNumPerPartitionPerNode = if (numPerPartition.nonEmpty) {
+      numPerPartitionPerNode + (index -> numPerPartition.get)
+    } else {
+      numPerPartitionPerNode
+    }
+
+    // Create the new state
+    optimizationState.copy(registers = newRegisters, numPerPartitionPerNode = newNumPerPartitionPerNode)
+  }
+
+  /**
+   * Optimizes the transformer being applied by the instruction at the given index.
+   * Outputs a new optimization state.
+   *
+   * @param os The previous optimization state
+   * @param transformerApplyNode
+   * @param index
+   * @return
+   */
+  private def optimizeTransformer(os: OptimizationState, transformerApplyNode: TransformerApplyNode, index: Int)
+  : OptimizationState = {
+    val tIndex = transformerApplyNode.transformer
+    val inputIndices = transformerApplyNode.inputs
+
+    val numPerPartition = os.numPerPartitionPerNode(inputIndices.head)
+    val inputs = inputIndices.map(os.registers).collect {
+      case RDDOutput(rdd) => rdd
+    }
+
+    // Optimize the transformer if possible
+    os.registers(tIndex) match {
+      case TransformerOutput(ot: OptimizableTransformer[a, b]) => {
```

Comment:
Is there a reason you're using lower-case type names here? Makes it a little surprising/hard to read.